Add support for using a custom docker image

fiam opened the pull request:
Add a -docker-image argument, which can be either an image name
or a Dockerfile. This allows the user to use a custom docker image
which is useful for e.g. compiling the Linux version when it has
additional dependencies.

fiam commented:
@lucor I'm planning to make a couple more PRs to allow me to build my app for Windows, Linux and macOS with fyne-cross, but I have some questions regarding the current design choices in the codease. Could you please let me know why you're only using `fyne package` for macOS? Windows and Linux seem to be built by calling `go build` instead.

lucor commented on cmd/fyne-cross/main.go:
```diff
@@ -79,6 +81,7 @@ func main() {
 	flag.Usage = printUsage
 
 	defaultTarget := strings.Join([]string{build.Default.GOOS, build.Default.GOARCH}, "/")
+	flag.StringVar(&dockerImage, "docker-image", builder.DefaultDockerImage, "Docker image to use. Can be an image name or the path to a Dockerfile")
```
we should figure out a way to specify a custom image for each target.
fyne-cross can accept multiple targets and in the feature we could split base images for each os/arch (see #25)

fiam replied:
I honestly see no need for having a per-target docker image, it would increase maintenance burden a lot. It would also require users to have a per target custom Dockerfile when they need to make changes. Invoking `fyne-cross` once per target seems a perfectly fine solution for me. Most of the time you will need to call target specific tools anyway (e.g. code signing for macOS and Windows).

lucor replied:
> I honestly see no need for having a per-target docker image, it would increase maintenance burden a lot.  It would also require users to have a per target custom Dockerfile when they need to make changes.

I agree that having a per-target image increase the maintenance burden but there are some other things we have to consider too. The fyne-cross objective is to support as much as possible all targets supported by Fyne. With that in mind along with the "desktop" image there is already a dedicated image for android we have to support. Additionally there are plans to support FreeBSD and iOS (although I don't know if this is even feasible right now :) ) that will have dedicated image too.

Least but not last #25 could allow to shrink image size and make it more usable even in zones when the internet connection speed is not high. Thinking of use cases where for example you are on darwin and want build only for another platform.

> Invoking fyne-cross once per target seems a perfectly fine solution for me. Most of the time you will need to call target specific tools anyway (e.g. code signing for macOS and Windows).

Fair enough, but in this case I'd prefer to slightly change the command line args/options (introducing a break change) and introduce something like sub commands to have it less error prone, like:
`fyne-cross <target> --arch=[] --docker-image="lucor/custom-target-image:latest"`

What do you think?

fiam replied:
> Fair enough, but in this case I'd prefer to slightly change the command line args/options (introducing a break change) and introduce something like sub commands to have it less error prone, like:
> `fyne-cross <target> --arch=[] --docker-image="lucor/custom-target-image:latest"`
> 
> What do you think?

I'm not sure how that's different than the `-docker-image="lucor/custom-target-image:latest"` flag implemented in this PR (except for the double dash). Could you please elaborate?

lucor replied:
The main point we need to address is how to have the possibility to specify a custom image per target (theoretically we'd allow also per arch but better leave this out now to avoid not needed over-engineering).

To achieve this some ways come in mind right now:
a) specify the custom image using a per os flag: `-docker-image-linux`, `-docker-image-windows`, `-docker-image-android`, ecc..
b) specify the custom image using an env variable: `FYNE_CROSS_DOCKER_IMAGE_LINUX`, `FYNE_CROSS_DOCKER_IMAGE_ANDROID`, ecc...
c) update fyne-cross to use subcommands: 
```
fyne-cross <OS> --docker-image="<custom_image>" #builds <OS>
fyne-cross linux --docker-image="<custom_image>" #builds linux
fyne-cross android --docker-image="<custom_image>" #builds android
```

Solution c) would introduce a breaking change but in the long term could be a good choice since every OS could require specific options.

With regard to the `-docker-image="lucor/custom-target-image:latest"` what I meant was related to the value of the image could be only defined as:
- `<repository>:<tag>` to use remote or local tagged images (i.e. fiam/fyne-cross-windows:latest)
- the image id of a local image: (i.e `5d70d40d6b7c`)

lucor commented on internal/builder/docker.go:
```diff
@@ -0,0 +1,130 @@
+	DefaultDockerImage = "lucor/fyne-cross:1.4.0"
+)
+
+type dockerImageProvider interface {
```
What about having only the Builder responsible for the definition of the custom docker image ?
Adding something like `DockerImage() string` and `SetDockerImage(string)` could be good enough?

lucor commented on internal/builder/docker.go:
```diff
@@ -0,0 +1,130 @@
+// image name provider and then calling Run on the returned command
+func runDockerCmd(vol *volume.Volume, env []string, workDir string, command []string, verbose bool) error {
+	// retrieve image name
+	image, err := dockerProvider.DockerImageName()
```
If possible, it is preferable to have all dependencies declared as function argument to allow test / mocking.

fiam replied:
This function seems to take already too many arguments. Plus, it's also very small (`dockerCmd()` does most of the work) and you can still test by changing the package-level docker image beforehand. What I hadn't consider is that at some point you might want to run different builds in parallel, and in that case package-level state definitely gets on the way.

lucor replied:
Yes, run builds in parallel could be something we'd want to add in future.
Anyway, since we are going to add the `DockerImage` method into the builder interface what about to use only the dockerCmd passing into it the image as b.DockerImage() ?
It won't reduce the number of arguments but will self document the required dependencies.

fiam replied:
With the latest commits all package level state should now be gone. All state is now encapsulated inside each Builder implementation.

lucor replied:
Indeed, but what I meant is that `dockerProvider` was not an explict dependency since it not passed as function argument :-)

lucor commented on internal/builder/docker.go:
```diff
@@ -0,0 +1,130 @@
+	Path string
+}
+
+func (d *dockerFile) DockerImageName() (string, error) {
```
Adding support for local image is a nice to have but this mean we need to add both formats for the docker image (remote and local). What about supporting only the remote one for now ?
The build of the local image can be handled with an external script.

fiam replied:
There's no need to add the Dockefile to the repo. I'm using this Dockerfile:

```
FROM lucor/fyne-cross:1.4.0
RUN apt-get update && apt-get -y install libglib2.0-dev libgtk-3-dev
```

This uses your image and just runs another command on top to install the dependencies.

fiam replied:
Just realised you meant running a local image by its name. I'll add support for that too.

lucor replied:
The idea was to support only remote. In this way fyne-cross will be only responsible to all task related to app build. What do you think ?

fiam replied:
Just to be clear, I'm considering 3 sources of images here:

- Image from name pulled from docker registry (the only supported way and hardcoded for now)
- Image from using an existing image in the system (Docker handles this transparently in the same way as the previous case)
- Image from a Dockerfile

I understand that what you are suggesting is dropping the support for providing a Dockerfile to simplify the code, and instead require the user build the image via their build scripts then pass just the image name to fyne-cross. I don't think that simplifies the code that much, and being able to pass just a Dockerfile and let fyne-cross set it up is way more convenient.

Consider that passing interfaces around is way more maintanable than explicit basic type parameters (strings, ints, bools, etc...) down the line, because you can extend the functions handling them without having to change all the signatures.

lucor replied:
I see that adding support for a Docker file and building the image inside fyne-cross could be useful.

But this would add an extra feature that we need to maintain or support. I'm thinking of all the problem could be related to building an image itself and not due to the `docker build -f /path/Dockerfile` command :-)

The feature can be addressed with an external step and as you correctly pointed before in any case most of the time you will need to call target specific tools anyway.

lucor reviewed (changes requested):
@fiam thanks for this PR, I agree that adding support for custom images would be helpful when the application needs additional dependencies. Added some thoughts.

Wrt your question:
> Could you please let me know why you're only using fyne package for macOS?

The fyne package is used only to package the application, all the binaries are compiled using `go build`. The only exception is windows where the windows resources needed to add icon and properties are generated using the windres tool.

Please also note that every build phase (as defined into the Builder interface) has three steps:
- PreBuild where all check about build requirements are done. Example `go mod init` or for Windows the creation of the windows resources that will be linked at compile time.
- Build where the app is compiled
- Package where the distribution package is created

lucor commented:
Could you also please update the PR to merge against the `develop` branch ?
Thanks!

fiam commented:
@lucor I'll be merging my changes into the develop branch and pushing the changes later today. Thanks for the review!

fiam committed:
Add support for using a custom docker image

Add a -docker-image argument, which can be either an image name
or a Dockerfile. This allows the user to use a custom docker image
which is useful for e.g. compiling the Linux version when it has
additional dependencies.

fiam committed:
Rename dockerRemoteImage to dockerImageName

Specifying a docker image name works for both local and remote
images without having to do any extra work, so the dockerRemoteImage
type name was confusing.

lucor commented on internal/builder/android.go:
```diff
@@ -30,7 +28,7 @@ func (b *Android) PreBuild(vol *volume.Volume, opts PreBuildOptions) error {
 		return fmt.Errorf("appID is required for android build")
 	}
-	return goModInit(vol, opts.Verbose)
+	return goModInit(b, vol, opts.Verbose)
```
```suggestion
	return goModInit(b.DockerImage(), vol, opts.Verbose)
```
What about passing only the docker image now that is defined the DockerImage method into the interface.

fiam replied:
Note that `DockerImage()` can return an error too. Passing the image name would require an error check at every call site. Passing the Builder seems like a cleaner design to me, since `goModInit()` won't even use the `Builder` nor the image name itself, it just passes it to `runBuilderDockerCmd()` which serves as an entry point of all docker related commands run by a builder. Also, at some point in the future you might want to check some attributes of the Builder in either `goModInit()` or `runBuilderDockerCmd()`. By passing the Builder you don't need to change the function signature.

lucor replied:
I was referring to the `DockerImage() string` method defined into the Builder interface that returns only a string.

Please don't get me wrong, I fully agree with you that passing an interface is a cleaner design.
What I'm saying is that in this case considering that since the code in not exposed (internal package) and that the method returns a string it could be good enough to left the `goModInit()` and `dockerCmd()` signatures as is accepting a string.

If we want to go anyway for a clean desing we'd also define for something like:

```
type DockerImager interface {
   DockerImage() string
}
```
embed the `DockerImager` into the `Builder` interface and then use the `DockerImager` into the `goModInit()` and `dockerCmd()` signatures. WDYT?

lucor commented:
@fiam thanks for all feedback and work. 
Sorry for all comments, but I think we're really close now :-)

Here some thoughts with the attempt to recap the main open points:
- we'd support a custom image per target. To define them in this phase I'd go for env variables or flags. I'd slightly prefer env variables over flags and let the sub command approach for a next release.
- we'd support only existing docker images, no support to build via Dockerfile. If we want to validate the source, supporting only the format `<repository>:<tag>` is ok as first step.

lucor commented:
Support for using a custom docker image has been added in v2.0.0-beta1.
@fiam thanks for helping shaping up this feature!
